src/main/java/org/traccar/protocol/LaipacProtocolDecoder.java

```diff
@@ -81,37 +99,114 @@ private String decodeAlarm(String event) {
                 return Position.ALARM_GEOFENCE_EXIT;
             case "6":
                 return Position.ALARM_OVERSPEED;
+            case "5":
+                return Position.ALARM_POWER_CUT;
             case "3":
                 return Position.ALARM_SOS;
             default:
                 return null;
         }
     }
 
-    @Override
-    protected Object decode(
-            Channel channel, SocketAddress remoteAddress, Object msg) throws Exception {
+    private String decodeEvent(String event, Position position) {
 
-        String sentence = (String) msg;
+        if (event.length() == 1) {
+            char inputStatus = event.charAt(0);
+            if (inputStatus >= 'A' && inputStatus <= 'D') {
+                int inputStatusInt = inputStatus - 'A';
+                position.set(Position.PREFIX_IN + 1, inputStatusInt & 1);
+                position.set(Position.PREFIX_IN + 2, inputStatusInt & 2);
+                position.set(Position.KEY_IGNITION, ((inputStatusInt & 1) != 0) ? true : false);
+                return null;
+            }
+        }
 
-        if (sentence.startsWith("$ECHK") && channel != null) {
-            channel.writeAndFlush(new NetworkMessage(sentence + "\r\n", remoteAddress)); // heartbeat
-            return null;
+        return event;
+    }
+
+    private String getDevicePassword(DeviceSession deviceSession) {
+
+        String devicePassword = defaultDevicePassword;
+
+        Device device = Context.getIdentityManager().getById(deviceSession.getDeviceId());
+        if (device != null) {
+            String password = device.getString(Command.KEY_DEVICE_PASSWORD);
+            if (password != null) {
+                devicePassword = password;
+            }
         }
 
-        Parser parser = new Parser(PATTERN, sentence);
-        if (!parser.matches()) {
-            return null;
+        return devicePassword;
+    }
+
+    private String getEventResponse(String event, String devicePassword) {
+
+        String responseCode = null;
+
+        switch (event) {
+            case "3":
+                responseCode = "d";
+                break;
+            case "S":
+            case "T":
+                responseCode = "t";
+                break;
+            case "X":
+            case "4":
+                responseCode = "x";
+                break;
+            case "Y":
+                responseCode = "y";
+                break;
+            case "Z":
+                responseCode = "z";
+                break;
+            default:
+                break;
+        }
+
+        if (responseCode != null) {
+            String response = "$AVCFG," + devicePassword + "," + responseCode;
+            response += Checksum.nmea(response) + "\r\n";
+            return response;
+        }
+
+        return null;
+    }
+
+    private String getStatusResponse(String status, String event, String checksum) {
```

Comment:
The common practice we have is to have a method that sends response instead of returning something. Please check other protocols for examples.

Reply:
Going that way... Thanks.